File: Projects/jc/Sources/jc/Primitives/StringTokenizer.cpp
/*
 * 작성자: 윤정도
 * 생성일: 7/7/2024 9:10:17 PM
 * =====================
 *
 */

#include "StringTokenizer.h"

NS_JC_BEGIN

//////////////////////////////////////////////////////////////////////////////////////////
StringTokenizer::StringTokenizer(const char* _str)
: pStr_(_str)
{
	strLen_ = StringUtil::Length(_str);
	delims_.Reserve(4);
}

//////////////////////////////////////////////////////////////////////////////////////////
StringTokenizer::StringTokenizer(const String& _str)
: pStr_(_str.Source())
{
	strLen_ = _str.Length();
	delims_.Reserve(4);
}

//////////////////////////////////////////////////////////////////////////////////////////
bool StringTokenizer::IsDelim(char _delim)
{
	return delims_.Exist(_delim);
}

//////////////////////////////////////////////////////////////////////////////////////////
void StringTokenizer::AddDelim(char _delim)
{
	if (IsDelim(_delim))
		return;

	delims_.PushBack(_delim);
}

//////////////////////////////////////////////////////////////////////////////////////////
bool StringTokenizer::HasNext()
{
	if (cur_ >= strLen_)
		return false;

	// 현재 위치부터 남은 문자열에서 비딜리미터가 있는지 확인
	for (int i = cur_; i < strLen_; ++i)
	{
		if (!IsDelim(pStr_[i]))
			return true;
	}
	return false;
}

//////////////////////////////////////////////////////////////////////////////////////////
String StringTokenizer::Next(OUT char* _catchedDelim /*= nullptr*/)
{
	// 연속된 딜리미터 스킵
	while (cur_ < strLen_ && IsDelim(pStr_[cur_]))
	{
		++cur_;
	}

	// 토큰 시작 위치 설정 (prev_ 업데이트)
	prev_ = cur_;

	// 토큰 끝 찾기
	for (; cur_ < strLen_; ++cur_)
	{
		char ch = pStr_[cur_];
		if (IsDelim(ch))
		{
			if (_catchedDelim)
			{
				*_catchedDelim = ch;
			}
			break;
		}
	}

	String ret = StringUtil::SubStr(pStr_, strLen_, prev_, cur_ - prev_);
	cur_ = cur_ + 1;	// 딜리미터 스킵
	return ret;
}


NS_END

File: Projects/SteinsGate-Tools.Common/Project/Base/StringTokenizer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace SGToolsCommon.Base
{
    /// <summary>
    /// 문자/문자열 구분자 기반 토크나이저. ParseDictionary 포함.
    /// </summary>
    public struct StringTokenizer
    {
        public const char INVALID_DELIM = '긁';

        public const int DELIM_TYPE_NONE   = -1;
        public const int DELIM_TYPE_CHAR   =  0;
        public const int DELIM_TYPE_STRING =  1;

        private readonly string str_;
        private readonly List<char> delims_;
        private List<string>? strDelims_;
        private int prev_;
        private int cur_;
        private int max_;

        public int Cur  => cur_;
        public int Prev => prev_;

        public int LastOffset = -1;
        public int LastDelimType = -1;
        public char LastDelim = INVALID_DELIM;
        public string LastDelimStr = string.Empty;

        public bool Option_ReturnStart  { get; set; } = false;
        public bool Option_ReturnString { get; set; } = true;

        public StringTokenizer(string str, int _max = int.MaxValue)
        {
            str_       = str ?? string.Empty;
            delims_    = new List<char>(4);
            strDelims_ = null;
            max_       = _max;
            prev_      = 0;
            cur_       = 0;
        }

        public bool IsDelim(char delim) => delims_.Contains(delim);

        public bool IsDelim(string delim)
        {
            if (string.IsNullOrEmpty(delim)) return false;
            if (delim.Length == 1) return IsDelim(delim[0]);
            if (strDelims_ == null) return false;
            return strDelims_.Contains(delim);
        }

        public bool IsEnd() => cur_ >= str_.Length;

        public void AddDelim(char delim)
        {
            if (!IsDelim(delim))
                delims_.Add(delim);
        }

        public void AddDelim(string delim)
        {
            if (string.IsNullOrEmpty(delim))
                throw new Exception("빈 문자열 전달하지마");
            if (delim.Length == 1)
                AddDelim(delim[0]);
            else
            {
                if (!IsDelim(delim))
                {
                    strDelims_ ??= new List<string>(4);
                    strDelims_.Add(delim);
                }
            }
        }

        public void RemoveDelim(char delim)
        {
            if (IsDelim(delim))
                delims_.Remove(delim);
        }

        public void SetCur(int _cur)   => cur_  = _cur;
        public void SetPrev(int _prev) => prev_ = _prev;

        public string? Next()
        {
            LastDelim    = INVALID_DELIM;
            LastDelimStr = string.Empty;
            LastOffset   = -1;
            LastDelimType = DELIM_TYPE_NONE;

            if (cur_ >= str_.Length)
                return null;

            int start = prev_;
            for (; cur_ < str_.Length && cur_ < max_; cur_++)
            {
                char ch = str_[cur_];
                if (IsDelim(ch))
                {
                    LastDelimType = DELIM_TYPE_CHAR;
                    LastDelim = ch;
                    break;
                }

                if (strDelims_ != null)
                {
                    foreach (string strDelim in strDelims_)
                    {
                        if (cur_ + strDelim.Length <= str_.Length &&
                            cur_ + strDelim.Length <= max_ &&
                            str_.Substring(cur_, strDelim.Length) == strDelim)
                        {
                            LastDelimType = DELIM_TYPE_STRING;
                            LastDelimStr  = strDelim;
                            cur_ += strDelim.Length - 1;
                            goto OUTER_BREAK;
                        }
                    }
                }
            }
        OUTER_BREAK:
            string ret = string.Empty;
            if (Option_ReturnString)
                ret = str_.Substring(start, cur_ - start);

            prev_ = cur_ + 1;
            cur_  = prev_;

            LastOffset = Option_ReturnStart ? start : cur_;
            return ret;
        }

        public static Dictionary<string, string> ParseDictionary(string _str)
        {
            var tokenizer = new StringTokenizer(_str);
            tokenizer.AddDelim(':');
            tokenizer.AddDelim(',');
            var dict = new Dictionary<string, string>();
            while (true)
            {
                string? key = tokenizer.Next();
                if (string.IsNullOrEmpty(key)) break;
                if (tokenizer.LastDelim != ':') break;
                key = key.Trim();
                string? value = tokenizer.Next();
                if (value == null || tokenizer.LastDelim != ',') break;
                value = value.Trim();
                dict.Add(key, value);
            }
            return dict;
        }
    }
}
